aws-datastore/src/main/java/com/amplifyframework/datastore/syncengine/Merger.kt

```diff
@@ -55,82 +52,103 @@ internal class Merger(
-                // and the version would get bumped up.
-                .filter { currentVersion: Int -> currentVersion == -1 || incomingVersion > currentVersion } // If we should merge, then do so now, starting with the model data.
-                .flatMapCompletable {
-                    val firstStep: Completable = if (mutationOutbox.hasPendingMutation(
-                            model.primaryKeyString,
-                            model.modelName
-                        )
-                    ) {
+        return rxCompletable {
+            // if no models to merge, return early
+            if (modelsWithMetadata.isEmpty()) {
+                return@rxCompletable
+            }
+
+            // create (key, model metadata) map for quick lookup to re-associate
+            val modelMetadataMap = modelsWithMetadata.associateBy({ it.syncMetadata.primaryKeyString }, { it })
+
+            // Consumer to announce Model merges
+            val mergedConsumer = Consumer<ModelWithMetadata<T>> {
+                announceSuccessfulMerge(it)
+                LOG.debug("Remote model update was sync'd down into local storage: $it")
+            }
+            // Reusable consumer to pass into each operation, which will broadcast change type to mergedConsumer
+            val modelChangeConsumer: Consumer<StorageItemChange<T>> = Consumer { changeTypeConsumer.accept(it.type()) }
+
+            // After we receive a report of a successful metadata item process, we notify the merge consumer
+            val metadataChangeConsumer: Consumer<StorageItemChange<ModelMetadata>> = Consumer { change ->
+                modelMetadataMap[change.item().primaryKeyString]?.let {
+                    mergedConsumer.accept(it)
+                }
+            }
+
+            // fetch a Map of all model versions from Metadata table
+            val localModelVersions = versionRepository.fetchModelVersions(modelsWithMetadata)
+
+            /*
+            Fetch a Set of all pending mutation ids for type T
+            We exclude in-flight mutations from the returned pending mutation. If a mutation is excluded, it is
+            likely that the subscription processor returned the item before the outbox response. We want to process
+            whichever comes first
+             */
+            val pendingMutations = mutationOutbox.fetchPendingMutations(
+                models = modelsWithMetadata.map { it.model },
+                modelClass = modelsWithMetadata.first().model.javaClass.name,
+                excludeInFlight = true
+            )
+
+            // Construct a batch of operations to be executed in a single transactions
+            val operations = modelsWithMetadata.mapNotNull {
+                val incomingVersion = it.syncMetadata.version ?: -1
+                val localVersion = localModelVersions.getOrDefault(
+                    "${it.model.modelName}|${it.model.primaryKeyString}",
```

Comment:
This string construction appears in a few places, at least here and in the version repository. Let's make and extension value on `Model` to construct this key and use that in place of the literal string constructions.

Reply:
done. Was very careful in how I named to variable to make sure its never accidentally used. This is only needed to match the way the sqlite key is constructed.